overlord/servicestate/servicemgr.go

```diff
@@ -42,8 +52,245 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
+	if err != nil {
+		return err
+	}
+
+	// if nothing was modified or we are not on UC18+, we are done
+	if len(modified) == 0 || deviceCtx.Classic() || deviceCtx.Model().Base() == "" {
+		m.ensuredSnapSvcs = true
+		return nil
+	}
+
+	// otherwise we need to check for all the services that were modified if
+	// they were recently killed when we modified usr-lib-snapd.mount as part of
+	// a snapd snap refresh
+
+	// as a last resort, if we fail in trying to start any services, we should
+	// reboot
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		// TODO: reboot the system immediately here
+	}()
+
+	// we decide on which services to restart by identifying (out of the set of
+	// services we just modified) services that were stopped after
+	// usr-lib-snapd.mount was written, but before usr-lib-snapd.mount was last
+	// stopped - this is the time window in which snapd (accidentally) killed
+	// all snap services using Requires=, see LP #1924805 for full details, so
+	// we need to undo that by restarting those snaps
+
+	// TODO: use the var from core18.go here instead
+	st, err := os.Stat("/etc/systemd/system/usr-lib-snapd.mount")
+	if err != nil {
+		return err
+	}
+
+	// TODO: we should check if usr-lib-snapd.mount was modified before the
+	// current boot time, if it was then we can just skip this since we know
+	// any service stops that happened were unrelated
+	lowerTimeBound := st.ModTime()
+
+	// connect to the system bus to talk to systemd
+	conn, err := dbus.SystemBus()
+	if err != nil {
+		panic(err)
+	}
+	defer conn.Close()
+
+	// TODO: don't hard-code this
+	usrLibSnapdMountUnitObjPath := dbus.ObjectPath("/org/freedesktop/systemd1/unit/usr_2dlib_2dsnapd_2emount")
+
+	// sanity check, dbus is hard
+	if !usrLibSnapdMountUnitObjPath.IsValid() {
+		return fmt.Errorf("internal error: usr-lib-snapd.mount dbus object path is wrong")
+	}
+
+	// connect to systemd with the usr-lib-snapd.mount unit object
+	busObj := conn.Object("org.freedesktop.systemd1", usrLibSnapdMountUnitObjPath)
+
+	// Get the InactiveEnterTimestamp property from the Unit interface, this is
+	// the time that usr-lib-snapd.mount was transitioned from deactivating to
+	// inactive and was done being started. This is the correct upper bound for
+	// our window in which systemd killed snap services because systemd order's
+	// the transactions when we stop usr-lib-snapd.mount thusly:
+	//
+	// 1. Find all units which have Requires=usr-lib-snapd.mount (all snap
+	//    services  which would have been refreshed during snapd 2.49.2)
+	// 2. Stop all such services found in 1.
+	// 3. Stop usr-lib-snapd.mount itself.
+	//
+	// Thus the time after all the services were killed is given by the time
+	// that systemd transitioned usr-lib-snapd.mount to inactive, which is given
+	// by InactiveEnterTimestamp.
+	res, err := busObj.GetProperty("org.freedesktop.systemd1.Unit.InactiveEnterTimestamp")
```

Comment:
isn't this the same as `systemctl show --property InactiveEnterTimestampMonotonic`?